// next@16.2.9/dist/shared/lib/router/utils/get-segment-param.js
import { __exports as getSegmentParam } from '../../../../../../../../../../_virtual/get-segment-param.js';
import { __require as requireInterceptionRoutes } from './interception-routes.js';

var hasRequiredGetSegmentParam;

function requireGetSegmentParam () {
	if (hasRequiredGetSegmentParam) return getSegmentParam;
	hasRequiredGetSegmentParam = 1;
	(function (exports) {
		Object.defineProperty(exports, "__esModule", {
		    value: true
		});
		function _export(target, all) {
		    for(var name in all)Object.defineProperty(target, name, {
		        enumerable: true,
		        get: all[name]
		    });
		}
		_export(exports, {
		    getParamProperties: function() {
		        return getParamProperties;
		    },
		    getSegmentParam: function() {
		        return getSegmentParam;
		    },
		    isCatchAll: function() {
		        return isCatchAll;
		    }
		});
		const _interceptionroutes = requireInterceptionRoutes();
		function getSegmentParam(segment) {
		    const interceptionMarker = _interceptionroutes.INTERCEPTION_ROUTE_MARKERS.find((marker)=>segment.startsWith(marker));
		    // if an interception marker is part of the path segment, we need to jump ahead
		    // to the relevant portion for param parsing
		    if (interceptionMarker) {
		        segment = segment.slice(interceptionMarker.length);
		    }
		    if (segment.startsWith('[[...') && segment.endsWith(']]')) {
		        return {
		            // TODO-APP: Optional catchall does not currently work with parallel routes,
		            // so for now aren't handling a potential interception marker.
		            paramType: 'optional-catchall',
		            paramName: segment.slice(5, -2)
		        };
		    }
		    if (segment.startsWith('[...') && segment.endsWith(']')) {
		        return {
		            paramType: interceptionMarker ? `catchall-intercepted-${interceptionMarker}` : 'catchall',
		            paramName: segment.slice(4, -1)
		        };
		    }
		    if (segment.startsWith('[') && segment.endsWith(']')) {
		        return {
		            paramType: interceptionMarker ? `dynamic-intercepted-${interceptionMarker}` : 'dynamic',
		            paramName: segment.slice(1, -1)
		        };
		    }
		    return null;
		}
		function isCatchAll(type) {
		    return type === 'catchall' || type === 'catchall-intercepted-(..)(..)' || type === 'catchall-intercepted-(.)' || type === 'catchall-intercepted-(..)' || type === 'catchall-intercepted-(...)' || type === 'optional-catchall';
		}
		function getParamProperties(paramType) {
		    let repeat = false;
		    let optional = false;
		    switch(paramType){
		        case 'catchall':
		        case 'catchall-intercepted-(..)(..)':
		        case 'catchall-intercepted-(.)':
		        case 'catchall-intercepted-(..)':
		        case 'catchall-intercepted-(...)':
		            repeat = true;
		            break;
		        case 'optional-catchall':
		            repeat = true;
		            optional = true;
		            break;
		    }
		    return {
		        repeat,
		        optional
		    };
		}

		
	} (getSegmentParam));
	return getSegmentParam;
}

export { requireGetSegmentParam as __require };
