// next@15.5.15/dist/shared/lib/head.js
import { __module as head } from '../../../../../../../../_virtual/head2.js';
import { __require as require_interop_require_default } from '../../../../../../@swc_helpers@0.5.15/node_modules/@swc/helpers/cjs/_interop_require_default.js';
import { __require as require_interop_require_wildcard } from '../../../../../../@swc_helpers@0.5.15/node_modules/@swc/helpers/cjs/_interop_require_wildcard.js';
import { __require as requireJsxRuntime } from '../../../../../../react@18.2.0/node_modules/react/jsx-runtime.js';
import React__default from 'react';
import { __require as requireSideEffect } from './side-effect.js';
import { __require as requireAmpContext_sharedRuntime } from './amp-context.shared-runtime.js';
import { __require as requireHeadManagerContext_sharedRuntime } from './head-manager-context.shared-runtime.js';
import { __require as requireAmpMode } from './amp-mode.js';
import { __require as requireWarnOnce } from './utils/warn-once.js';

var hasRequiredHead;

function requireHead () {
	if (hasRequiredHead) return head.exports;
	hasRequiredHead = 1;
	(function (module, exports$1) {
		'use client';
		Object.defineProperty(exports$1, "__esModule", {
		    value: true
		});
		function _export(target, all) {
		    for(var name in all)Object.defineProperty(target, name, {
		        enumerable: true,
		        get: all[name]
		    });
		}
		_export(exports$1, {
		    default: function() {
		        return _default;
		    },
		    defaultHead: function() {
		        return defaultHead;
		    }
		});
		const _interop_require_default = /*@__PURE__*/ require_interop_require_default();
		const _interop_require_wildcard = /*@__PURE__*/ require_interop_require_wildcard();
		const _jsxruntime = requireJsxRuntime();
		const _react = /*#__PURE__*/ _interop_require_wildcard._(React__default);
		const _sideeffect = /*#__PURE__*/ _interop_require_default._(requireSideEffect());
		const _ampcontextsharedruntime = requireAmpContext_sharedRuntime();
		const _headmanagercontextsharedruntime = requireHeadManagerContext_sharedRuntime();
		const _ampmode = requireAmpMode();
		const _warnonce = requireWarnOnce();
		function defaultHead(inAmpMode) {
		    if (inAmpMode === void 0) inAmpMode = false;
		    const head = [
		        /*#__PURE__*/ (0, _jsxruntime.jsx)("meta", {
		            charSet: "utf-8"
		        }, "charset")
		    ];
		    if (!inAmpMode) {
		        head.push(/*#__PURE__*/ (0, _jsxruntime.jsx)("meta", {
		            name: "viewport",
		            content: "width=device-width"
		        }, "viewport"));
		    }
		    return head;
		}
		function onlyReactElement(list, child) {
		    // React children can be "string" or "number" in this case we ignore them for backwards compat
		    if (typeof child === 'string' || typeof child === 'number') {
		        return list;
		    }
		    // Adds support for React.Fragment
		    if (child.type === _react.default.Fragment) {
		        return list.concat(// @ts-expect-error @types/react does not remove fragments but this could also return ReactPortal[]
		        _react.default.Children.toArray(child.props.children).reduce(// @ts-expect-error @types/react does not remove fragments but this could also return ReactPortal[]
		        (fragmentList, fragmentChild)=>{
		            if (typeof fragmentChild === 'string' || typeof fragmentChild === 'number') {
		                return fragmentList;
		            }
		            return fragmentList.concat(fragmentChild);
		        }, []));
		    }
		    return list.concat(child);
		}
		const METATYPES = [
		    'name',
		    'httpEquiv',
		    'charSet',
		    'itemProp'
		];
		/*
		 returns a function for filtering head child elements
		 which shouldn't be duplicated, like <title/>
		 Also adds support for deduplicated `key` properties
		*/ function unique() {
		    const keys = new Set();
		    const tags = new Set();
		    const metaTypes = new Set();
		    const metaCategories = {};
		    return (h)=>{
		        let isUnique = true;
		        let hasKey = false;
		        if (h.key && typeof h.key !== 'number' && h.key.indexOf('$') > 0) {
		            hasKey = true;
		            const key = h.key.slice(h.key.indexOf('$') + 1);
		            if (keys.has(key)) {
		                isUnique = false;
		            } else {
		                keys.add(key);
		            }
		        }
		        // eslint-disable-next-line default-case
		        switch(h.type){
		            case 'title':
		            case 'base':
		                if (tags.has(h.type)) {
		                    isUnique = false;
		                } else {
		                    tags.add(h.type);
		                }
		                break;
		            case 'meta':
		                for(let i = 0, len = METATYPES.length; i < len; i++){
		                    const metatype = METATYPES[i];
		                    if (!h.props.hasOwnProperty(metatype)) continue;
		                    if (metatype === 'charSet') {
		                        if (metaTypes.has(metatype)) {
		                            isUnique = false;
		                        } else {
		                            metaTypes.add(metatype);
		                        }
		                    } else {
		                        const category = h.props[metatype];
		                        const categories = metaCategories[metatype] || new Set();
		                        if ((metatype !== 'name' || !hasKey) && categories.has(category)) {
		                            isUnique = false;
		                        } else {
		                            categories.add(category);
		                            metaCategories[metatype] = categories;
		                        }
		                    }
		                }
		                break;
		        }
		        return isUnique;
		    };
		}
		/**
		 *
		 * @param headChildrenElements List of children of <Head>
		 */ function reduceComponents(headChildrenElements, props) {
		    const { inAmpMode } = props;
		    return headChildrenElements.reduce(onlyReactElement, []).reverse().concat(defaultHead(inAmpMode).reverse()).filter(unique()).reverse().map((c, i)=>{
		        const key = c.key || i;
		        if (process.env.NODE_ENV === 'development') {
		            // omit JSON-LD structured data snippets from the warning
		            if (c.type === 'script' && c.props['type'] !== 'application/ld+json') {
		                const srcMessage = c.props['src'] ? '<script> tag with src="' + c.props['src'] + '"' : "inline <script>";
		                (0, _warnonce.warnOnce)("Do not add <script> tags using next/head (see " + srcMessage + "). Use next/script instead. \nSee more info here: https://nextjs.org/docs/messages/no-script-tags-in-head-component");
		            } else if (c.type === 'link' && c.props['rel'] === 'stylesheet') {
		                (0, _warnonce.warnOnce)('Do not add stylesheets using next/head (see <link rel="stylesheet"> tag with href="' + c.props['href'] + '"). Use Document instead. \nSee more info here: https://nextjs.org/docs/messages/no-stylesheets-in-head-component');
		            }
		        }
		        return /*#__PURE__*/ _react.default.cloneElement(c, {
		            key
		        });
		    });
		}
		/**
		 * This component injects elements to `<head>` of your page.
		 * To avoid duplicated `tags` in `<head>` you can use the `key` property, which will make sure every tag is only rendered once.
		 */ function Head(param) {
		    let { children } = param;
		    const ampState = (0, _react.useContext)(_ampcontextsharedruntime.AmpStateContext);
		    const headManager = (0, _react.useContext)(_headmanagercontextsharedruntime.HeadManagerContext);
		    return /*#__PURE__*/ (0, _jsxruntime.jsx)(_sideeffect.default, {
		        reduceComponentsToState: reduceComponents,
		        headManager: headManager,
		        inAmpMode: (0, _ampmode.isInAmpMode)(ampState),
		        children: children
		    });
		}
		const _default = Head;

		if ((typeof exports$1.default === 'function' || (typeof exports$1.default === 'object' && exports$1.default !== null)) && typeof exports$1.default.__esModule === 'undefined') {
		  Object.defineProperty(exports$1.default, '__esModule', { value: true });
		  Object.assign(exports$1.default, exports$1);
		  module.exports = exports$1.default;
		}

		
	} (head, head.exports));
	return head.exports;
}

export { requireHead as __require };
