// next@15.5.15/dist/client/set-attributes-from-props.js
import { __module as setAttributesFromProps } from '../../../../../../../_virtual/set-attributes-from-props.js';

var hasRequiredSetAttributesFromProps;

function requireSetAttributesFromProps () {
	if (hasRequiredSetAttributesFromProps) return setAttributesFromProps.exports;
	hasRequiredSetAttributesFromProps = 1;
	(function (module, exports$1) {
		Object.defineProperty(exports$1, "__esModule", {
		    value: true
		});
		Object.defineProperty(exports$1, "setAttributesFromProps", {
		    enumerable: true,
		    get: function() {
		        return setAttributesFromProps;
		    }
		});
		const DOMAttributeNames = {
		    acceptCharset: 'accept-charset',
		    className: 'class',
		    htmlFor: 'for',
		    httpEquiv: 'http-equiv',
		    noModule: 'noModule'
		};
		const ignoreProps = [
		    'onLoad',
		    'onReady',
		    'dangerouslySetInnerHTML',
		    'children',
		    'onError',
		    'strategy',
		    'stylesheets'
		];
		function isBooleanScriptAttribute(attr) {
		    return [
		        'async',
		        'defer',
		        'noModule'
		    ].includes(attr);
		}
		function setAttributesFromProps(el, props) {
		    for (const [p, value] of Object.entries(props)){
		        if (!props.hasOwnProperty(p)) continue;
		        if (ignoreProps.includes(p)) continue;
		        // we don't render undefined props to the DOM
		        if (value === undefined) {
		            continue;
		        }
		        const attr = DOMAttributeNames[p] || p.toLowerCase();
		        if (el.tagName === 'SCRIPT' && isBooleanScriptAttribute(attr)) {
		            el[attr] = !!value;
		        } else {
		            el.setAttribute(attr, String(value));
		        }
		        // Remove falsy non-zero boolean attributes so they are correctly interpreted
		        // (e.g. if we set them to false, this coerces to the string "false", which the browser interprets as true)
		        if (value === false || el.tagName === 'SCRIPT' && isBooleanScriptAttribute(attr) && (!value || value === 'false')) {
		            // Call setAttribute before, as we need to set and unset the attribute to override force async:
		            // https://html.spec.whatwg.org/multipage/scripting.html#script-force-async
		            el.setAttribute(attr, '');
		            el.removeAttribute(attr);
		        }
		    }
		}

		if ((typeof exports$1.default === 'function' || (typeof exports$1.default === 'object' && exports$1.default !== null)) && typeof exports$1.default.__esModule === 'undefined') {
		  Object.defineProperty(exports$1.default, '__esModule', { value: true });
		  Object.assign(exports$1.default, exports$1);
		  module.exports = exports$1.default;
		}

		
	} (setAttributesFromProps, setAttributesFromProps.exports));
	return setAttributesFromProps.exports;
}

export { requireSetAttributesFromProps as __require };
