// next@15.5.15/dist/shared/lib/router/utils/route-match-utils.js
import { __exports as routeMatchUtils } from '../../../../../../../../../../_virtual/route-match-utils.js';
import { __require as requirePathToRegexp } from '../../../../compiled/path-to-regexp/index.js';
import { __require as requireRoutePatternNormalizer } from '../../../../lib/route-pattern-normalizer.js';

/**
 * Client-safe utilities for route matching that don't import server-side
 * utilities to avoid bundling issues with Turbopack
 */

var hasRequiredRouteMatchUtils;

function requireRouteMatchUtils () {
	if (hasRequiredRouteMatchUtils) return routeMatchUtils;
	hasRequiredRouteMatchUtils = 1;
	(function (exports$1) {
		Object.defineProperty(exports$1, "__esModule", {
		    value: true
		});
		function _export(target, all) {
		    for(var name in all)Object.defineProperty(target, name, {
		        enumerable: true,
		        get: all[name]
		    });
		}
		_export(exports$1, {
		    safeCompile: function() {
		        return safeCompile;
		    },
		    safePathToRegexp: function() {
		        return safePathToRegexp;
		    },
		    safeRegexpToFunction: function() {
		        return safeRegexpToFunction;
		    },
		    safeRouteMatcher: function() {
		        return safeRouteMatcher;
		    }
		});
		const _pathtoregexp = requirePathToRegexp();
		const _routepatternnormalizer = requireRoutePatternNormalizer();
		function safePathToRegexp(route, keys, options) {
		    if (typeof route !== 'string') {
		        return (0, _pathtoregexp.pathToRegexp)(route, keys, options);
		    }
		    // Check if normalization is needed and cache the result
		    const needsNormalization = (0, _routepatternnormalizer.hasAdjacentParameterIssues)(route);
		    const routeToUse = needsNormalization ? (0, _routepatternnormalizer.normalizeAdjacentParameters)(route) : route;
		    try {
		        return (0, _pathtoregexp.pathToRegexp)(routeToUse, keys, options);
		    } catch (error) {
		        // Only try normalization if we haven't already normalized
		        if (!needsNormalization) {
		            try {
		                const normalizedRoute = (0, _routepatternnormalizer.normalizeAdjacentParameters)(route);
		                return (0, _pathtoregexp.pathToRegexp)(normalizedRoute, keys, options);
		            } catch (retryError) {
		                // If that doesn't work, fall back to original error
		                throw error;
		            }
		        }
		        throw error;
		    }
		}
		function safeCompile(route, options) {
		    // Check if normalization is needed and cache the result
		    const needsNormalization = (0, _routepatternnormalizer.hasAdjacentParameterIssues)(route);
		    const routeToUse = needsNormalization ? (0, _routepatternnormalizer.normalizeAdjacentParameters)(route) : route;
		    try {
		        return (0, _pathtoregexp.compile)(routeToUse, options);
		    } catch (error) {
		        // Only try normalization if we haven't already normalized
		        if (!needsNormalization) {
		            try {
		                const normalizedRoute = (0, _routepatternnormalizer.normalizeAdjacentParameters)(route);
		                return (0, _pathtoregexp.compile)(normalizedRoute, options);
		            } catch (retryError) {
		                // If that doesn't work, fall back to original error
		                throw error;
		            }
		        }
		        throw error;
		    }
		}
		function safeRegexpToFunction(regexp, keys) {
		    const originalMatcher = (0, _pathtoregexp.regexpToFunction)(regexp, keys || []);
		    return (pathname)=>{
		        const result = originalMatcher(pathname);
		        if (!result) return false;
		        // Clean parameters before returning
		        return {
		            ...result,
		            params: (0, _routepatternnormalizer.stripParameterSeparators)(result.params)
		        };
		    };
		}
		function safeRouteMatcher(matcherFn) {
		    return (pathname)=>{
		        const result = matcherFn(pathname);
		        if (!result) return false;
		        // Clean parameters before returning
		        return (0, _routepatternnormalizer.stripParameterSeparators)(result);
		    };
		}

		
	} (routeMatchUtils));
	return routeMatchUtils;
}

export { requireRouteMatchUtils as __require };
